### PR TITLE
add unix-over-ssh transport

### PR DESCRIPTION
**Pull request by AkihiroSuda**
# EDIT: need design review https://github.com/docker/go-connections/pull/39#issuecomment-312765226

- - - 

part of `docker -H ssh://host` (https://github.com/moby/moby/pull/33566)

**How I did it**

- on Unix: By `os/exec`-uting `ssh -L local.sock:/var/run/docker.sock`. So it should even work with `~/.ssh/config`.

- on Windows: using pageant + Go-native ssh (`golang.org/x/crypto/ssh`)

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

### CODE REVIEW

**Comment by tonistiigi** [comment]
Usually, programs that do this call into itself through the ssh connection instead of creating a temporary socket that may create a security problem. For example `rsync --server`, `scp -t`, `systemd-stdio-bridge` etc. There is also `ssh -W` but not sure if it works for unix sockets. Not completely against the current method but something to consider. I guess the advantage of this is that it works in new client-old daemon combinations but not sure how important that is.

**Comment by AkihiroSuda** [comment]
Can we use `ssh $host socat STDIO UNIX-CONNECT:/var/run/docker.sock`, although `socat` is rarely installed on the remote host?

**Comment by AkihiroSuda** [comment]
Or `nc`

**Comment by AkihiroSuda** [comment]
On second thought it might be good to put socat/nc functionality into `docker/cli` for ease of security configuration.

i.e. `docker sockhelper` connects stdio to `/var/run/docker.sock` or `//.pipe/docker-engine`.
(This subcommand should not be shown in `docker --help`)

WDYT @docker/core-cli-maintainers

**Comment by AkihiroSuda** [comment]
any thought?

**Comment by thaJeztah** [comment]
ping @ijc @tianon @diogomonica - any thoughts?

File: sockets/sockets.go [hunk 1]
```diff
@@ -36,3 +40,28 @@ func ConfigureTransport(tr *http.Transport, proto, addr string) error {
...
+}
+
+// DialSSH connects to a Unix socket over SSH.
+func DialSSH(addr string) (net.Conn, error) {
+	u, err := url.Parse("ssh://" + addr)
+	if err != nil {
+		return nil, err
+	}
+	if u.User == nil || u.User.Username() == "" {
+		return nil, fmt.Errorf("ssh requires username")
```
Comment (ijc) [inline sockets/sockets.go]: Normally ssh defaults to the current username.

File: sockets/ssh_windows.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+func dialSSH(user, addr, socketPath string) (net.Conn, error) {
+	if !pageant.Available() {
+		return nil, errors.New("pageant not running")
+	}
+	auth := ssh.PublicKeysCallback(pageant.New().Signers)
+	sshClient, err := ssh.Dial("tcp", addr, &ssh.ClientConfig{
+		User:            user,
+		Auth:            []ssh.AuthMethod{auth},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
```
Comment (ijc) [inline sockets/ssh_windows.go]: The [docs](https://github.com/golang/crypto/blob/master/ssh/client.go#L229..L231) say:
```
// InsecureIgnoreHostKey returns a function that can be used for
// ClientConfig.HostKeyCallback to accept any host key. It should
// not be used for production code.
```
I think that is sound advice.

**Comment by ijc** [comment]
I left a couple of implementation comments, but I'm not too sure about the functionality generally. Is this not something which can be achieved with the use of ssh's `DynamicForward` builtin SOCKS proxy?

A more general solution might be to support executing a proxy command, similar to ssh's `ProxyCommand` which is commonly combined with `DynamicForward` to produce transparent tunnels.

**Comment by justincormack** [comment]
I don't think we should make any assumption that the `docker` cli is installed on the remote machine, now that it is a standalone repo not part of Moby project. I can see a use case for a special command that can be connected to an ssh submodule and does the equivalent of `docker sockhelper` so that it is easy to configure `sshd` to just support the Docker use case but not shell logins, but this is quite an advanced use case.

**Comment by tonistiigi** [comment]
> I don't think we should make any assumption that the docker cli is installed on the remote machine

It could be `dockerd sockhelper` as it wouldn't need anything from cli and only dials into the socket.

**Comment by justincormack** [comment]
SSH subsystems have to be a single binary, you can't specify arguments.

**Comment by tonistiigi** [comment]
I didn't mean making the subsystem but calling it directly like most other apps(except sftp). Although, the manual says "Arguments should be a subsystem name and a command (with optional arguments) to execute upon subsystem request".

**Comment by justincormack** [comment]
If the subsystem was dockerd then you could call it with any arguments,
sure but that's no use for access control.

The point of a subsystem is to be able to provide users only controlled
socket access without any other access. Obviously you would need authz to
make this actually secure.

**Comment by tianon** [comment]
I usually use something like `ssh -L localhost:2375:/var/run/docker.sock ...` manually when I need this sort of thing (since SSH's port forwarding natively supports unix sockets, as mentioned above).  I also often run that in a container with `0.0.0.0` instead of `localhost`, so that I can connect directly to `2375` on the container's IP to hit a remote Docker engine (and let my local Docker engine take care of reconnecting via `--restart=always`, cleanup, keeping the access local-only, etc).

When I saw the PR title (before looking at the implementation), I thought this might be using [`golang.org/x/crypto/ssh`](https://godoc.org/golang.org/x/crypto/ssh) to pipe through the SSH connection directly from Go, which I'm -1 on for a user-facing feature since I think if we invoke anything related to the SSH name, users will expect their local `.ssh/config` to take effect and thus be surprised when things like `ssh://my-ssh-alias` doesn't work and apply their local `ProxyCommand` settings.  So if we do provide this functionality, I think we'll have to continue shelling out to make it respect `.ssh/config` and user expectations, although I'm still dubious, especially given my experiences doing similar things by hand.

I've run into all sorts of fun edge cases with TTYs and shared connections that I think we'll eventually have to grow warts like `-S none` to handle.

**Comment by AkihiroSuda** [comment]
Method | No socket file leak | Support `~/.ssh/config` | No dependency on Docker product | Likely to work on default Docker installation
--- | --- | --- | --- | --- 
`golang.org/x/crypto/ssh` | :white_check_mark: | :x:  | :white_check_mark: | :white_check_mark: |
`golang.org/x/crypto/ssh`+implement `~/.ssh/config` parser | :white_check_mark: | :white_check_mark: (hard)  | :white_check_mark: | :white_check_mark: |
`os/exec("ssh -L local.sock:/var/run/docker.sock)` | :x: | :white_check_mark:  | :white_check_mark: | :white_check_mark: |
`os/exec("ssh socat-or-nc")` | :white_check_mark: | :white_check_mark:  | :white_check_mark: | :x: (`socat` and `nc` are rarelly installed) |
`os/exec("ssh docker sockproxy")` | :white_check_mark: | :white_check_mark:  | :x: | :white_check_mark: |
`os/exec("ssh dockerd sockproxy")` | :white_check_mark: | :white_check_mark:  | :x: (`dockerd` binary name might change on Moby Project) | :white_check_mark: |
`os/exec("ssh docker sockproxy \|\| ssh socat-or-nc")` | :white_check_mark: | :white_check_mark:  | :white_check_mark: | :white_check_mark: |
SSH subsystem + `os/exec("ssh -s somesubsystem")` | :white_check_mark: | :white_check_mark:  | :white_check_mark: | :x: (needs to modify `sshd_config`) |

I'm -1 on subsystem because it is unlikely to work on default Docker installation.
How about `os/exec("ssh docker sockproxy || ssh socat-or-nc")`?

Also, `golang.org/x/crypto/ssh`+implementing  `~/.ssh/config` parser might be still attractive option, though it would be hard.
(It would be easy if we can depend on `libssh`, but I think we should not introduce C dependency)

**Comment by AkihiroSuda** [comment]
InfraKit seems to have implement similar thing using golang SSH
https://github.com/docker/infrakit/pull/654

cc @chungers

Can we move this forward?

**Comment by justincormack** [comment]
Getting the golang implementation to parse hashed known_hosts files seems
to be difficult too, I discovered yesterday.

**Comment by AkihiroSuda** [comment]
`os/exec("ssh docker sockproxy || ssh socat-or-nc")`  seems better then?

**Comment by thaJeztah** [comment]
ping @justincormack @tianon @ijc ^^

File: sockets/ssh_unix.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	if host == "" {
+		return nil, fmt.Errorf("invalid host: %q", host)
+	}
+	return []string{
+		ssh,
+		"-L", localSocketPath + ":" + remoteSocketPath,
+		"-N",
+		"-l", user,
+		"-p", strconv.Itoa(port),
+		host,
```
Comment (ijc) [inline sockets/ssh_unix.go]: It'd be worth comparing this with what options e.g. `git` or `rsync` use. It seems like `git` uses `-batch` which might be interesting.

**Comment by ijc** [comment]
@thaJeztah I responded to you [previous ping](https://github.com/docker/go-connections/pull/39#issuecomment-311246961) with a few comments:
* https://github.com/docker/go-connections/pull/39#discussion_r124228144
* https://github.com/docker/go-connections/pull/39#discussion_r124228696
* https://github.com/docker/go-connections/pull/39#issuecomment-311309758
Doesn't look like there has been a response to any of those. I made one new one just now as well.

**Comment by AkihiroSuda** [comment]
@ijc Thank you for reviewing, but the current status is not code review. (sorry I didn't make it clear here)

We need to determine the design before fixing up the code.

I listed possible designs in https://github.com/docker/go-connections/pull/39#issuecomment-312765226 .
The PR implements  the `os/exec("ssh -L local.sock:/var/run/docker.sock)` strategy, but it appeared to  look like incorrect for security reason. (socket file leak)

So I'm currently proposing the `os/exec("ssh docker sockproxy || ssh socat-or-nc")` strategy instead, and waiting for SGTMs.

## Pseudo-code of the `os/exec("ssh docker sockproxy || ssh socat-or-nc")` strategy 
```go
import "os/exec"
var cmd *exec.Cmd
if exec.Command("ssh", host, "sh", "-c", "which docker").Run() != nil {
		// This requires server-side Docker CLI to implement "sockproxy" subcommand, which connects stdio to /var/run/docker.sock.
		cmd = exec.Command("ssh", host, "docker", "sockproxy")
		// Note that Docker CLI might no be installed on 3rd party Moby distribution.  So we also support nc & socat as well.
		// (Note that nc & socat are rarely installed on the server, so we lookup `docker sockproxy` by default.)
} else if exec.Command("ssh", host, "sh", "-c", "which nc").Run() != nil {
		// Note: busybox nc has different options. So we also need to detect busybox nc. (omitted in this pseudo-code)
		cmd = exec.Command("ssh", host, "nc", "-U", "/var/run/docker.sock")
} else if exec.Command("ssh", host, "sh", "-c", "which socat").Run() != nil {
		cmd = exec.Command("ssh", host, "socat, "stdio", "unix-connect:/var/run/docker.sock")
}
connectToStdio(cmd)

```

**Comment by AkihiroSuda** [comment]
closing for now: https://github.com/moby/moby/pull/33566#issuecomment-356634154

**Pull request closed without merging**